pkg/syncer/upsync/upsync_reconcile.go

```diff
@@ -0,0 +1,261 @@
+		// => let's update the resource upstream.
+		logger.Info("Updating upstream resource")
+		preparedResource := c.prepareResourceForUpstream(ctx, gvr, upstreamNamespace, upstreamClusterName, downstreamResource)
+		if err != nil {
+			return false, err
+		}
+
+		if !dirtyStatus {
+			// if no status needs to be upsynced upstream, then we can set the resource version annotation to the downstream resource version
+			// at the same time as we update the main resource content.
+
+			preparedResource.SetAnnotations(addResourceVersionAnnotation(resourceVersionDownstream, preparedResource.GetAnnotations()))
+			existingResource, err := c.upstreamClient.Resource(gvr).Cluster(upstreamClusterName.Path()).Namespace(upstreamNamespace).Get(ctx, preparedResource.GetName(), metav1.GetOptions{})
+			if err != nil {
+				return false, err
+			}
+			preparedResource.SetResourceVersion(existingResource.GetResourceVersion())
+			_, err = c.upstreamClient.Resource(gvr).Cluster(upstreamClusterName.Path()).Namespace(upstreamNamespace).Update(ctx, preparedResource, metav1.UpdateOptions{})
+			// If the downstream resource is marked for deletion, let's requeue it to manage the deletion timestamp
+			return markedForDeletionDownstream, err
+		}
+
+		// Status also needs to be upsynced so let's do it in 3 steps:
+		// - update the upstream resource main content
+
+		preparedResource.SetAnnotations(addResourceVersionAnnotation(resourceVersionUpstream, preparedResource.GetAnnotations()))
+		existingResource, err := c.upstreamClient.Resource(gvr).Cluster(upstreamClusterName.Path()).Namespace(upstreamNamespace).Get(ctx, preparedResource.GetName(), metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		preparedResource.SetResourceVersion(existingResource.GetResourceVersion())
+		updatedResource, err := c.upstreamClient.Resource(gvr).Cluster(upstreamClusterName.Path()).Namespace(upstreamNamespace).Update(ctx, preparedResource, metav1.UpdateOptions{})
+		if err != nil {
+			return false, err
+		}
+
+		// - update the status as a distinct action,
+		preparedResource.SetResourceVersion(updatedResource.GetResourceVersion())
+		updatedResource, err = c.upstreamClient.Resource(gvr).Cluster(upstreamClusterName.Path()).Namespace(upstreamNamespace).UpdateStatus(ctx, preparedResource, metav1.UpdateOptions{})
+		if err != nil {
+			return false, err
+		}
+
+		// - finally update the main content again to set the resource version annotation to the value of the downstream resource version.
+		preparedResource.SetAnnotations(addResourceVersionAnnotation(resourceVersionDownstream, preparedResource.GetAnnotations()))
+		preparedResource.SetResourceVersion(updatedResource.GetResourceVersion())
+		_, err = c.upstreamClient.Resource(gvr).Cluster(upstreamClusterName.Path()).Namespace(upstreamNamespace).Update(ctx, preparedResource, metav1.UpdateOptions{})
+		// If the downstream resource is marked for deletion, let's requeue it to manage the deletion timestamp
```

Comment:
why not fall through?

Reply:
It was so at start. But stopping there and using requeue avoid having a bit `if/else` block for both cases (quick path / slow path): we return directly at the end of the quick path and requeue.
So I kept the "return with requeue" at the end of the slow path for consistency.
But I can move back to the previous structure, sure.

Reply:
k